src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRule.scala

```diff
@@ -154,12 +151,41 @@ object JoinIndexRule
     // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
     // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
     // lead to wrong results. So we only replace the base relation.
-    plan transform {
-      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+
+    val replacedPlan = plan transformUp {
+      case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
         val updatedOutput =
           baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
         baseRelation.copy(relation = relation, output = updatedOutput)
     }
+
+    val complementIndexPlan = plan transformUp {
+      case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+
+        val files =
+          fsRelation.location.inputFiles.map(new Path(_))
+            .filter(!index.allSourceFileSet.contains(_))
+        val newLocation =
+          new InMemoryFileIndex(spark, files, Map(), None)
+        val newRelation =
+          fsRelation.copy(location = newLocation, dataSchema = index.schema)(spark)
+
+        val newLogicalRelation = baseRelation.copy(relation = newRelation, output = updatedOutput)
+
+        val attrs = newLogicalRelation.attributeMap.baseMap
+          .filter { key =>
+            index.indexedColumns contains key._2._2.name
+          }
+          .map { case (_, ref) => ref._1 }
+        // TODO fix
+        //  cur: project => filter => exchange => relation
+        //  desired: exchange => project => filter => relation
```

Comment:
I see. I thought the exchange was pushed so that it can be reused, by looking at the example in the description. :)

Reply:
Yes I think push down filter/project might be more efficient than reuse. What do you think??